### DISCUSSION TITLE
Svelte 4.0 soon to be released

### DISCUSSION THREAD

**Discussion by jamestagal** [plentico/plenti#281] [General]
I see from the Svelte dev log that Svelte version 4 might be released this month. It sounds like mainly a maintenance release with no real breaking changes... They do talk about vastly improved '''custom elements''' and I wondered whether you made use of them in Plenti @jimafisk ... and whether this would be a welcomed improvement.?

See the discussion about the changes to custom elements from their [Dev Log May 2023](https://www.youtube.com/live/aDTc_uFufS8?feature=share&t=1415) on YouTube.

**Comment by jimafisk** [comment]
> Svelte version 4 might be released this month

Oh wow that came up fast!

> whether this would be a welcomed improvement.?

Definitely, we've been stuck on an older version of Svelte for a bit now. There is an upgrade path by using esbuild to convert the compiler to an older version of JS so it can run in V8 with less preprocessing. I'll circle back on this so we can be prepared.

> See the discussion about the changes to custom elements from their [Dev Log May 2023](https://www.youtube.com/live/aDTc_uFufS8?feature=share&t=1415) on YouTube.

Thanks for sending the link, I hadn't seen this!

**Comment by jamestagal** [comment]
Hi @jimafisk,
Nice to see this release v0.6.13 with compiler and Svelte updates. https://github.com/plentico/plenti/releases/tag/v0.6.13
What were the main changes?

BTW. You were beaten by the Svelte 4.0 release by one hour. 😃 

<img width="897" alt="github" src="https://github.com/plentico/plenti/assets/6309595/119119fb-ceee-4d24-8816-4ae707fda8f5">

**Reply by jimafisk** [reply]
Ha! I didn't even realize, that's great that `v4.0.0` is here :tada:. 

> What were the main changes?

I'm not entirely sure, but there are likely features that you couldn't do with the older version. I usually stick to the basics so I don't know exactly what's been added. I did notice when upgrading a project that it's stricter about re-assigning `const` variables, which aren't supposed to be reassigned. We typically add something like `thing = thing` to ensure reactivity updates, but now you just need to make sure you initially define `let thing` not `const thing` if you want to do that.

I'll try the new compiler transformation process on `4.0.0`, hopefully we can pull that in without breaking everything!

**Comment by jamestagal** [comment]
Hi all,

Here is the official blog post of [Svelte 4.0's stable release](https://svelte.dev/blog/svelte-4) 😸 

Enjoy!

**Reply by jimafisk** [reply]
Thanks for staying on top of this @jamestagal! 

> Svelte 4 reduces the Svelte package size by nearly 75%

That's very exciting, I think most of that reduction in dependencies (reduced from 61 down to 16) but if the compiler has shrunk at all that can help some of the V8 burden.

> Svelte 5 will be a rewrite of the Svelte compiler and runtime...Svelte 5 will bring major new features and performance improvements to Svelte.

This is very interesting, I wonder if they are considering move to a compiled language like Rust or Zig. That would be great from a portability standpoint for us.

**Reply by jamestagal** [reply]
Yes it's all very exciting. They seem to be not too far away from sharing more information on Svelte 5 as well. 
Are you going to give Svelte 4 a spin with Plenti? I did see a migration guide linked in the post too. 

I do some A11y error messages since the last Svelte update...see a previous message..https://github.com/plentico/plenti/issues/261#issuecomment-1580785426

**Reply by jimafisk** [reply]
I just attempted giving v4 a try, I'm confused with the claim that dependencies were "reduced from 61 down to 16." From what I can tell, Svelte `v3.x` doesn't have any dependencies, and it has 40 devDependencies. Svelte `v4.x` still has 19 devDependencies but adds 13 dependencies that now get added to the project. I'm not really keen on adding 13 new packages that have to ship with Plenti that we didn't have before. In an ideal world we'd be removing `node_modules` entirely from our project scaffolding.

**Reply by jamestagal** [reply]
Oh that puts a dampener on the excitement. It might be worth holding off to do more testing in terms of performance? What are the options/ alternatives?
